### ISSUE TITLE
Transfer Learning - Freezing Parameters

### ISSUE THREAD

**Issue by ska6845** [ultralytics/yolov5#679]
## Does yolov5 support transfer learning?

While training models, is there a possibility to use pretrained weights and modify last few layers?

**Comment by glenn-jocher** [comment]
@shubhamag01 this is the default behavior when a pretrained model is specified:
`python train.py --weights yolov5s.pt`

**Comment by ska6845** [comment]
@glenn-jocher is it possible to train,  removing last few layers and using pretrained model and add new layers and train freezing rest of untouched layers

**Comment by glenn-jocher** [comment]
@shubhamag01 you can do whatever you want

**Comment by karen-gishyan** [comment]
@glenn-jocher  is there a difference between 

python3 train.py --data coco1cls.data --cfg yolov3-spp.cfg --weights weights/yolov3-spp.pt 

and 

python3 train.py --data coco1cls.data --cfg yolov3-spp.cfg --weights weights/yolov3-spp.pt --transfer 

I assumed that  pre-trained weights was the idea behind transfer learning, then I found the tutorial on Transfer Learning it with  --transfer command specified.  
Thanks in advance.

**Comment by glenn-jocher** [comment]
@karen-gishyan your argument does not exist in train.py. See the argparser arguments at the end of train.py for a list of available arguments.

**Comment by karen-gishyan** [comment]
thanks @glenn-jocher.

**Comment by ska6845** [comment]
@glenn-jocher I had made a custom yolov5 model and i ran  `python train.py --img 640 --batch 16 --epochs 100 --data '../data.yaml' --cfg ./models/custom_yolov5s.yaml --weights yolov5s.pt --nosave --cache`, I have modified last few layers of yolov5s.yaml and made custom_yolov5s.yaml . Now I want the freeze the layers which are not being modified for yolov5s and want to train my model on remaining. How do I do that?

**Comment by glenn-jocher** [comment]
@ska6845 I've been asked this multiple times, so I've added a section to train.py that handles freezing parameters:
https://github.com/ultralytics/yolov5/blob/e71fd0ec0bbccb2ea38936a86defb0a31ff1153c/train.py#L76-L83

You can add any parameters you want to this list, with full or partial names, to freeze them before training starts. This code freezes all weights, leaving only biases with active gradients:

```python
    # Freeze
    model.info()
    freeze = ['.weight', ]  # parameter names to freeze (full or partial)
    if any(freeze):
        for k, v in model.named_parameters():
            if any(x in k for x in freeze):
                print('freezing %s' % k)
                v.requires_grad = False
    model.info()
```

Output:
```python
Model Summary: 191 layers, 7.46816e+06 parameters, 7.46816e+06 gradients
freezing model.0.conv.conv.weight
freezing model.0.conv.bn.weight
freezing model.1.conv.weight
freezing model.1.bn.weight
...
Model Summary: 191 layers, 7.46816e+06 parameters, 11453 gradients
```

**Comment by glenn-jocher** [comment]
@glenn-jocher TODO: update this to act correctly with optimizer parameter grouping (pg0-gp2):
https://github.com/ultralytics/yolov5/blob/e71fd0ec0bbccb2ea38936a86defb0a31ff1153c/train.py#L89-L91

**Comment by glenn-jocher** [comment]
Removing TODO as this fix is incorporated in PR #1239. 

Layer freezing functionality now operates correctly in all cases. To freeze layers, simply add their names to the `freeze` list in train.py:
https://github.com/ultralytics/yolov5/blob/187f7c2ed16b9eb9754b28e7b0aa397f908155aa/train.py#L83-L90

**Comment by pngmafia** [comment]
@glenn-jocher I wanted to freeze the backbone part of yolov5l configuration. Could you please tell me how to do it? And also will freezing the layers help in any way except decreased time in training. I am using coco pretrain for the logo-detection problem.
Thanks

**Comment by glenn-jocher** [comment]
@pngmafia freezing layers will reduce your mAP. You can add the names of the parameters you'd like to freeze to the `freeze` list. You can verify which layers are frozen by printing the model info:
```python
print(model.info(verbose=True))
```

**Comment by pngmafia** [comment]
@glenn-jocher I don't mind a decrease in mAP. But I need recall to be high. And also is there a way to give recall more weight than mAP?
Thanks

**Comment by glenn-jocher** [comment]
@pngmafia recall is not a universal metric, it depends on your conf. If you want maximum recall, all you need to do is set conf_thres to zero. Then you will have 100% recall.

**Comment by pngmafia** [comment]
@glenn-jocher can't I use the fitness function in utils. Use different weights for those four metrics? P R mAP:.5 and mAP:.5:.95. I see its 0 0 0.1 and 0.9 now.

**Comment by glenn-jocher** [comment]
@pngmafia sure, you can customize the hyperparameter evolution fitness function as you see fit. See hyperparameter evolution tutorial in https://docs.ultralytics.com/yolov5

**Comment by pngmafia** [comment]
@glenn-jocher Does changing those weights to 0 0.8 0.1 0.1 give me better recall? As compared to 0 0 0.1 and 0.9. Considering I use the same confidence threshold for both the experiments.

**Comment by glenn-jocher** [comment]
@pngmafia hyperparameter evolution maximizes the fitness function here:
https://github.com/ultralytics/yolov5/blob/187f7c2ed16b9eb9754b28e7b0aa397f908155aa/utils/general.py#L926-L930

Normal training minimizes loss on your training dataset, and is unrelated to hyperparameter evolution.

**Comment by pngmafia** [comment]
@glenn-jocher so this fitness function is just used only when we train with --evolve option. If I'm training normally on my dataset then, It really doesn't matter what weights I give in that function right?

**Comment by glenn-jocher** [comment]
@pngmafia that's correct.

**Comment by vedal** [comment]
@glenn-jocher thanks for including these changes to freeze params. 
How would you suggest going about freezing all but the last layer using your code, as would be done in a classical transfer learning setting (as suggested in https://docs.ultralytics.com/guides/model-training-tips/)? Which layers would you avoid freezing in this case?

**Comment by glenn-jocher** [comment]
@vedal see **Transfer Learning with Frozen Layers** tutorial below:

## YOLOv5 Tutorials

* [Train Custom Data](https://docs.ultralytics.com/yolov5/tutorials/train_custom_data)&nbsp; 🚀 RECOMMENDED
* [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/)&nbsp; ☘️ RECOMMENDED
* [Weights & Biases Logging](https://github.com/ultralytics/yolov5/issues/1289)&nbsp; 🌟 NEW
* [Supervisely Ecosystem](https://docs.ultralytics.com/yolov5/tutorials/model_export8)&nbsp; 🌟 NEW
* [Multi-GPU Training](https://docs.ultralytics.com/yolov5/tutorials/multi_gpu_training)
* [PyTorch Hub](https://docs.ultralytics.com/yolov5/tutorials/pytorch_hub_model_loading)&nbsp; ⭐ NEW
* [TorchScript, ONNX, CoreML Export](https://docs.ultralytics.com/yolov5/tutorials/model_export) 🚀
* [Test-Time Augmentation (TTA)](https://docs.ultralytics.com/yolov5/tutorials/test_time_augmentation)
* [Model Ensembling](https://docs.ultralytics.com/yolov5/tutorials/model_ensembling)
* [Model Pruning/Sparsity](https://docs.ultralytics.com/yolov5/tutorials/model_pruning_and_sparsity)
* [Hyperparameter Evolution](https://docs.ultralytics.com/yolov5/tutorials/hyperparameter_evolution)
* [Transfer Learning with Frozen Layers](https://docs.ultralytics.com/yolov5/tutorials/transfer_learning_with_frozen_layers)&nbsp; ⭐ NEW
* [TensorRT Deployment](https://github.com/wang-xinyu/tensorrtx)

**Comment by vedal** [comment]
@glenn-jocher cant believe I missed this! Thanks alot! :)
